src/modules/vtol_att_control/pusher.cpp

```diff
@@ -0,0 +1,319 @@
+
+#define ARSP_BLEND_START 8.0f	// airspeed at which we start blending mc/fw controls
+
+Pusher::Pusher(VtolAttitudeControl *attc) :
+	VtolType(attc),
+	_flag_enable_mc_motors(true),
+	_pusher_throttle(0.0f),
+	_mc_att_ctl_weight(1.0f)
+{
+	_vtol_schedule.flight_mode = MC_MODE;
+	_vtol_schedule.transition_start = 0;
+
+	_params_handles_pusher.front_trans_dur = param_find("VT_F_TRANS_DUR");
+	_params_handles_pusher.back_trans_dur = param_find("VT_B_TRANS_DUR");
+	_params_handles_pusher.pusher_mc = param_find("VT_TILT_MC");
+	_params_handles_pusher.pusher_transition = param_find("VT_TILT_TRANS");
+	_params_handles_pusher.pusher_fw = param_find("VT_TILT_FW");
+	_params_handles_pusher.airspeed_trans = param_find("VT_ARSP_TRANS");
+	_params_handles_pusher.elevons_mc_lock = param_find("VT_ELEV_MC_LOCK");
+ }
+
+Pusher::~Pusher()
+{
+}
+
+int
+Pusher::parameters_update()
+{
+	float v;
+	int l;
+
+	/* vtol duration of a front transition */
+	param_get(_params_handles_pusher.front_trans_dur, &v);
+	_params_pusher.front_trans_dur = math::constrain(v, 1.0f, 5.0f);
+
+	/* vtol duration of a back transition */
+	param_get(_params_handles_pusher.back_trans_dur, &v);
+	_params_pusher.back_trans_dur = math::constrain(v, 0.0f, 5.0f);
+
+	/* vtol pusher mechanism position in mc mode */
+	param_get(_params_handles_pusher.pusher_mc, &v);
+	_params_pusher.pusher_mc = v;
+
+	/* vtol pusher mechanism position in transition mode */
+	param_get(_params_handles_pusher.pusher_transition, &v);
+	_params_pusher.pusher_transition = v;
+
+	/* vtol pusher mechanism position in fw mode */
+	param_get(_params_handles_pusher.pusher_fw, &v);
+	_params_pusher.pusher_fw = v;
+
+	/* vtol airspeed at which it is ok to switch to fw mode */
+	param_get(_params_handles_pusher.airspeed_trans, &v);
+	_params_pusher.airspeed_trans = v;
+
+	/* vtol lock elevons in multicopter */
+	param_get(_params_handles_pusher.elevons_mc_lock, &l);
+	_params_pusher.elevons_mc_lock = l;
+
+	return OK;
+}
+
+void Pusher::update_vtol_state()
+{
+ 	parameters_update();
+
+ 	/* Simple logic using a two way switch to perform transitions.
+	 * After flipping the switch the vehicle will start the pusher (or tractor) motor, picking up
+	 * forward speed. After the vehicle has picked up enough speed the rotors shutdown. 
+	 * For the back transition the pusher motor is immediately stopped and rotors reactivated.
+ 	 */
+
+	if (_manual_control_sp->aux1 < 0.0f) {
+		// the transition to fw mode switch is off
+		if (_vtol_schedule.flight_mode == MC_MODE) {
+			// in mc mode
+			_vtol_schedule.flight_mode = MC_MODE;
+			_mc_att_ctl_weight = 1.0f;
+
+		} else if (_vtol_schedule.flight_mode == FW_MODE) {
+			// transition to mc mode
+			_vtol_schedule.flight_mode = TRANSITION_TO_MC;
+			_flag_enable_mc_motors = true;
+			_vtol_schedule.transition_start = hrt_absolute_time();
+
+		} else if (_vtol_schedule.flight_mode == TRANSITION_TO_FW) {
+			// failsafe back to mc mode
+			_vtol_schedule.flight_mode = MC_MODE;
```

Comment:
Added. It would be set 100ms later but I guess that is an eternity in a real-time world.